activerecord/lib/active_record/counter_cache.rb

```diff
@@ -43,23 +46,38 @@ def reset_counters(id, *counters, touch: nil)
             has_many_association = has_many_association.through_reflection
           end
 
+          counter_association = counter_association.to_sym
           foreign_key  = has_many_association.foreign_key.to_s
           child_class  = has_many_association.klass
           reflection   = child_class._reflections.values.find { |e| e.belongs_to? && e.foreign_key.to_s == foreign_key && e.options[:counter_cache].present? }
           counter_name = reflection.counter_cache_column
 
-          updates[counter_name] = object.send(counter_association).count(:all)
+          counts =
+            unscoped
+              .joins(counter_association)
+              .where(arel_table[primary_key].in(ids))
+              .group(arel_table[primary_key])
+              .count(:all)
```

Comment:
Do we need to use Arel here? Would something like this work instead?
```suggestion
              .joins(counter_association)
              .where(primary_key => ids)
              .group(primary_key)
              .count(:all)
```